In [1]:
# Section 1: Import Libraries
import pandas as pd  # For data manipulation
import numpy as np  # For numerical operations
import os  # For operating system interactions

In [3]:
# Define the path to the new dataset 
project_folder_path = r"C:\Users\marci\15-07-2024Instacart Basket Analysis\02 Data"

# STEP 1

In [5]:
# Import orders_products_combined.pkl into a DataFrame
df_ords_prods_merge = pd.read_pickle(r'C:\Users\marci\15-07-2024Instacart Basket Analysis\02 Data\Prepared Data\ords_prods_merge.pkl')

In [8]:
# Check for missing values in the combined DataFrame
missing_df_ords_prods_merge = df_ords_prods_merge.isnull().sum()
print("Missing values in combined DataFrame:")
print(missing_df_ords_prods_merge)

Missing values in combined DataFrame:
Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
product_id                0
add_to_cart_order         0
reordered                 0
existing_merge            0
product_name              0
aisle_id                  0
department_id             0
prices                    0
merge_flag                0
busiest_days              0
busiest_period_of_day     0
dtype: int64


In [10]:
df_ords_prods_merge.head()

Unnamed: 0  order_id  user_id  order_number  order_dow  order_hour_of_day  \
0           0   2539329        1             1          2                  8   
1           0   2539329        1             1          2                  8   
2           0   2539329        1             1          2                  8   
3           0   2539329        1             1          2                  8   
4           0   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   

  existing_merge                             product_name  aisle_id  \
0           both                                     Soda        77   
1           both  Organic Unsweetened Vanilla Almond Milk        91   
2           both                      Original Beef Jerky        23   
3           both               Aged White Cheddar Popcorn        23   
4           both         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices merge_flag    busiest_days busiest_period_of_day  
0              7     9.0       both  Regularly busy        Average orders  
1             16    12.5       both  Regularly busy        Average orders  
2             19     4.4       both  Regularly busy        Average orders  
3             19     4.7       both  Regularly busy        Average orders  
4             17     1.0       both  Regularly busy        Average orders

# STEP 2 :R epeating this process for the entire dataframe

In [11]:
# create a subset 
df = df_ords_prods_merge[:1000000]

In [13]:
df.shape

(1000000, 19)

# Aggregation Data with agg subset df

In [11]:
df.groupby('department_id').agg({'order_number': ['mean']})

NameError: name 'df' is not defined

#  Aggregation Data with agg /  entire dataset (df_ords_prods_merge)

In [12]:

df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.215751
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170371
12               15.887622
13               16.583304
14               16.759763
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [14]:
#Performing Multiple Aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

NameError: name 'df' is not defined

In [15]:
# with the entire dataframe
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457687   1  99
2                17.277920   1  99
3                17.179756   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225773   1  99
8                15.340520   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170371   1  99
12               15.887622   1  99
13               16.583304   1  99
14               16.759763   1  99
15               16.165037   1  99
16               17.663250   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

# STEP 3:Analyze the result

Comparison between: the subset (df), which represents 1 million records, and the entire data frame (df_ords_prods_merge) contains 32,433,030 records. 
1.Mean Order Numbers:
    The mean order numbers in the entire data frame are slightly higher than in the subset for most departments.
    For instance, Department 1 has a mean of 14.80 in the subset vs. 15.46 in the entire data frame.
2. Minimum /Maximum:
    The minimum order number is consistently 1 across all departments in both the subset and the entire data frame.
    The maximum order number is generally 99 across all departments in both data sets, with a few exceptions in the subset (e.g., Department 21 has a max of 97).
3. Variation in Departments:
    Departments 10 and 21 have higher mean order numbers in both data sets, indicating these departments might be more frequently ordered from.
    Departments like 1 and 9 show lower mean order numbers, which might indicate less frequent orders or smaller customer bases.

# STEP 4: Creating a loyalty flag for existing customers

In [16]:
#Determine the Maximum Order Number for Each Customer:
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform('max')

In [18]:
df_ords_prods_merge.head()

Unnamed: 0  order_id  user_id  order_number  order_dow  order_hour_of_day  \
0           0   2539329        1             1          2                  8   
1           0   2539329        1             1          2                  8   
2           0   2539329        1             1          2                  8   
3           0   2539329        1             1          2                  8   
4           0   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   

  existing_merge                             product_name  aisle_id  \
0           both                                     Soda        77   
1           both  Organic Unsweetened Vanilla Almond Milk        91   
2           both                      Original Beef Jerky        23   
3           both               Aged White Cheddar Popcorn        23   
4           both         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices merge_flag    busiest_days busiest_period_of_day  \
0              7     9.0       both  Regularly busy        Average orders   
1             16    12.5       both  Regularly busy        Average orders   
2             19     4.4       both  Regularly busy        Average orders   
3             19     4.7       both  Regularly busy        Average orders   
4             17     1.0       both  Regularly busy        Average orders   

   max_order  
0         10  
1         10  
2         10  
3         10  
4         10

In [20]:
# Create the Loyalty Flag Based on the Maximum Order Number:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [22]:
df_ords_prods_merge['loyalty_flag'].value_counts(dropna =False)


loyalty_flag
Regular customer    15890458
Loyal customer      10293434
New customer         6249138
Name: count, dtype: int64

In [33]:
 df_ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(100) 

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             1
2         1      New customer             1
3         1      New customer             1
4         1      New customer             1
..      ...               ...           ...
95        2  Regular customer             4
96        2  Regular customer             5
97        2  Regular customer             5
98        2  Regular customer             5
99        2  Regular customer             5

[100 rows x 3 columns]

# STEP 5: Determining whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [24]:
# Group by the loyalty_flag and calculate the basic statistics
loyalty_stats = df_ords_prods_merge.groupby('loyalty_flag')['prices'].agg(['mean', 'median', 'min', 'max']).reset_index()
print(loyalty_stats)

       loyalty_flag       mean  median  min      max
0    Loyal customer  10.388854     7.4  1.0  99999.0
1      New customer  13.294634     7.4  1.0  99999.0
2  Regular customer  12.496419     7.4  1.0  99999.0


# Analysis of the results: 
Loyal Customers:Tend to spend less on average compared to regular and new customers. This could indicate that loyal customers might be more price-conscious or prefer products with lower prices.
Regular Customers:Spend moderately compared to new and loyal customers. They may not yet be fully committed to the brand but are more engaged than new customers.
New Customers:Have the highest average spending, which might be due to exploratory behavior, trying out new products without regard to cost

# STEP 6: This will create a column spending_flag in entire DataFrame, categorizing each user as either a "Low spender" or a "High spender" based on their average spending.


In [26]:
# Calculate the mean prices per user
user_avg_prices = df_ords_prods_merge.groupby('user_id')['prices'].mean().reset_index()

# Create a new column for the spending flag
user_avg_prices['spending_flag'] = user_avg_prices['prices'].apply(lambda x: 'Low spender' if x < 10 else 'High spender')

In [28]:
# Merge the spending flag back into the original dataframe
df_ords_prods_merge = df_ords_prods_merge.merge(user_avg_prices[['user_id', 'spending_flag']], on='user_id', how='left')

# Verify the first few rows
df_ords_prods_merge.head()

Unnamed: 0  order_id  user_id  order_number  order_dow  order_hour_of_day  \
0           0   2539329        1             1          2                  8   
1           0   2539329        1             1          2                  8   
2           0   2539329        1             1          2                  8   
3           0   2539329        1             1          2                  8   
4           0   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     0.0         196                  1          0  ...   
1                     0.0       14084                  2          0  ...   
2                     0.0       12427                  3          0  ...   
3                     0.0       26088                  4          0  ...   
4                     0.0       26405                  5          0  ...   

                              product_name aisle_id  department_id  prices  \
0                                     Soda       77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk       91             16    12.5   
2                      Original Beef Jerky       23             19     4.4   
3               Aged White Cheddar Popcorn       23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls       54             17     1.0   

   merge_flag    busiest_days busiest_period_of_day max_order  loyalty_flag  \
0        both  Regularly busy        Average orders        10  New customer   
1        both  Regularly busy        Average orders        10  New customer   
2        both  Regularly busy        Average orders        10  New customer   
3        both  Regularly busy        Average orders        10  New customer   
4        both  Regularly busy        Average orders        10  New customer   

  spending_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  

[5 rows x 21 columns]

# STEP 7: Creating an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

In [30]:
# Calculate the median days since prior order per user
user_median_days = df_ords_prods_merge.groupby('user_id')['days_since_prior_order'].median().reset_index()

# Create a new column for the order frequency flag
user_median_days['order_frequency_flag'] = user_median_days['days_since_prior_order'].apply(
    lambda x: 'Non-frequent customer' if x > 20 else ('Regular customer' if 10 < x <= 20 else 'Frequent customer')
)

# Merge the order frequency flag back into the original dataframe
df_ords_prods_merge = df_ords_prods_merge.merge(user_median_days[['user_id', 'order_frequency_flag']], on='user_id', how='left')

# Verify the first few rows
df_ords_prods_merge.head()


Unnamed: 0  order_id  user_id  order_number  order_dow  order_hour_of_day  \
0           0   2539329        1             1          2                  8   
1           0   2539329        1             1          2                  8   
2           0   2539329        1             1          2                  8   
3           0   2539329        1             1          2                  8   
4           0   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     0.0         196                  1          0  ...   
1                     0.0       14084                  2          0  ...   
2                     0.0       12427                  3          0  ...   
3                     0.0       26088                  4          0  ...   
4                     0.0       26405                  5          0  ...   

  aisle_id department_id  prices  merge_flag    busiest_days  \
0       77             7     9.0        both  Regularly busy   
1       91            16    12.5        both  Regularly busy   
2       23            19     4.4        both  Regularly busy   
3       23            19     4.7        both  Regularly busy   
4       54            17     1.0        both  Regularly busy   

  busiest_period_of_day max_order  loyalty_flag  spending_flag  \
0        Average orders        10  New customer    Low spender   
1        Average orders        10  New customer    Low spender   
2        Average orders        10  New customer    Low spender   
3        Average orders        10  New customer    Low spender   
4        Average orders        10  New customer    Low spender   

  order_frequency_flag  
0     Regular customer  
1     Regular customer  
2     Regular customer  
3     Regular customer  
4     Regular customer  

[5 rows x 22 columns]

In [43]:
# Check for missing values in the combined DataFrame
missing_ords_prods_merge = df_ords_prods_merge.isnull().sum()
print("Missing values in combined DataFrame:")
print(missing_ords_prods_merge)

Missing values in combined DataFrame:
order_id                        0
user_id                         0
eval_set                        0
order_number                    0
order_dow                       0
order_hour_of_day               0
days_since_prior_order    2077979
product_id                      0
add_to_cart_order               0
reordered                       0
existing_merge                  0
product_name                28171
aisle_id                        0
department_id                   0
prices                          0
merge_flag                      0
Busiest day                     0
busiest_days                    0
busiest_period_of_day           0
max_order                       0
loyalty_flag                    0
spending_flag_x                 0
order_frequency_flag_x          0
spending_flag_y                 0
order_frequency_flag_y          0
dtype: int64


In [37]:
# checking Missing values in combined DataFrame:
# Check for missing values in the combined DataFrame
missing_df_ords_prods_merge = df_ords_prods_merge.isnull().sum()
print("Missing values in combined DataFrame:")
print(missing_df_ords_prods_merge)    
4

Missing values in combined DataFrame:
Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
product_id                0
add_to_cart_order         0
reordered                 0
existing_merge            0
product_name              0
aisle_id                  0
department_id             0
prices                    0
merge_flag                0
busiest_days              0
busiest_period_of_day     0
max_order                 0
loyalty_flag              0
spending_flag             0
order_frequency_flag      0
dtype: int64


4

# STEP 9: Exporting dataframe as a pickle file

In [39]:
df_ords_prods_merge.to_pickle(r"C:\Users\marci\15-07-2024Instacart Basket Analysis\02 Data\Prepared Data\ords_prods_merge.pkl")